File: apps/api/internal/inventory/service.go
package inventory

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"math"
	"sort"
	"strings"
	"sync"
	"time"
	"unicode/utf8"
)

type Service struct {
	repository Repository
	now        func() time.Time
	newID      func() string
	mutationMu sync.Mutex
}

type ServiceOption func(*Service)

func WithClock(now func() time.Time) ServiceOption {
	return func(service *Service) { service.now = now }
}

func WithIDGenerator(generator func() string) ServiceOption {
	return func(service *Service) { service.newID = generator }
}

func NewService(repository Repository, options ...ServiceOption) *Service {
	service := &Service{
		repository: repository,
		now:        func() time.Time { return time.Now().UTC() },
		newID:      randomID,
	}
	for _, option := range options {
		option(service)
	}
	return service
}

func (s *Service) ListItems(ctx context.Context, filter Filter) ([]Item, error) {
	if !filter.Archived.Valid() {
		return nil, ValidationError{Field: "archived", Message: "must be only or include"}
	}
	items, err := s.repository.ListItems(ctx, filter)
	if err != nil {
		return nil, err
	}

	now := s.now()
	for index := range items {
		events, eventErr := s.repository.ListEvents(ctx, items[index].ID, now.Add(-90*24*time.Hour))
		if eventErr != nil {
			return nil, eventErr
		}
		enrichItem(&items[index], events, now)
	}
	return items, nil
}

func (s *Service) Ping(ctx context.Context) error { return s.repository.Ping(ctx) }

func (s *Service) GetItem(ctx context.Context, id string) (Item, error) {
	item, err := s.repository.GetItem(ctx, strings.TrimSpace(id))
	if err != nil {
		return Item{}, err
	}
	events, err := s.repository.ListEvents(ctx, item.ID, s.now().Add(-90*24*time.Hour))
	if err != nil {
		return Item{}, err
	}
	enrichItem(&item, events, s.now())
	return item, nil
}

func (s *Service) CreateItem(ctx context.Context, input CreateItemInput) (Item, error) {
	name := strings.TrimSpace(input.Name)
	if name == "" {
		return Item{}, ValidationError{Field: "name", Message: "enter an item name"}
	}
	if utf8.RuneCountInString(name) > 120 {
		return Item{}, ValidationError{Field: "name", Message: "must be 120 characters or fewer"}
	}
	if input.Quantity < 0 {
		return Item{}, ValidationError{Field: "quantity", Message: "must be zero or greater"}
	}
	if input.MinQuantity < 0 {
		return Item{}, ValidationError{Field: "minQuantity", Message: "must be zero or greater"}
	}
	alternativeNames, err := normalizeValues("alternativeNames", input.AlternativeNames, name, 8, 120)
	if err != nil {
		return Item{}, err
	}
	categoryValues := input.Categories
	if len(categoryValues) == 0 && strings.TrimSpace(input.Category) != "" {
		categoryValues = []string{input.Category}
	}
	if len(categoryValues) == 0 {
		categoryValues = []string{"Other"}
	}
	categories, err := normalizeValues("categories", categoryValues, "", 9, 60)
	if err != nil {
		return Item{}, err
	}
	if len(categories) == 0 {
		categories = []string{"Other"}
	}

	mode := input.TrackingMode
	if mode == "" {
		mode = TrackingSimple
	}
	if !mode.Valid() {
		return Item{}, ValidationError{Field: "trackingMode", Message: "must be simple or exact"}
	}

	level := input.StockLevel
	if level != "" && !level.Valid() {
		return Item{}, ValidationError{Field: "stockLevel", Message: "must be full, okay, low, or out"}
	}
	levelPercent := 0.0
	if mode == TrackingExact {
		level = exactLevel(input.Quantity, input.MinQuantity)
	} else {
		if input.LevelPercent != nil {
			levelPercent = *input.LevelPercent
		} else if level.Valid() {
			levelPercent = percentForLevel(level)
		} else {
			levelPercent = 50
		}
		if !validPercent(levelPercent) {
			return Item{}, ValidationError{Field: "levelPercent", Message: "must be between zero and 100"}
		}
		level = simpleLevel(levelPercent)
	}

	location := fallback(strings.TrimSpace(input.Location), "Unassigned")
	if utf8.RuneCountInString(location) > 80 {
		return Item{}, ValidationError{Field: "location", Message: "must be 80 characters or fewer"}
	}
	unit := fallback(strings.TrimSpace(input.Unit), "item")
	if utf8.RuneCountInString(unit) > 30 {
		return Item{}, ValidationError{Field: "unit", Message: "must be 30 characters or fewer"}
	}
	now := s.now()
	item := Item{
		ID:               s.newID(),
		Name:             name,
		AlternativeNames: alternativeNames,
		Category:         categories[0],
		Categories:       categories,
		Location:         location,
		Unit:             unit,
		TrackingMode:     mode,
		Quantity:         input.Quantity,
		StockLevel:       level,
		LevelPercent:     levelPercent,
		MinQuantity:      input.MinQuantity,
		CreatedAt:        now,
		UpdatedAt:        now,
	}
	return s.repository.CreateItem(ctx, item)
}

func (s *Service) UpdateItem(ctx context.Context, itemID string, input UpdateItemInput) (Item, error) {
	s.mutationMu.Lock()
	defer s.mutationMu.Unlock()

	item, err := s.repository.GetItem(ctx, strings.TrimSpace(itemID))
	if err != nil {
		return Item{}, err
	}

	next := item
	if input.Name != nil {
		next.Name = strings.TrimSpace(*input.Name)
		if next.Name == "" {
			return Item{}, ValidationError{Field: "name", Message: "enter an item name"}
		}
		if utf8.RuneCountInString(next.Name) > 120 {
			return Item{}, ValidationError{Field: "name", Message: "must be 120 characters or fewer"}
		}
	}
	if input.AlternativeNames != nil {
		next.AlternativeNames, err = normalizeValues("alternativeNames", *input.AlternativeNames, next.Name, 8, 120)
		if err != nil {
			return Item{}, err
		}
	} else if input.Name != nil {
		next.AlternativeNames, err = normalizeValues("alternativeNames", next.AlternativeNames, next.Name, 8, 120)
		if err != nil {
			return Item{}, err
		}
	}
	if input.Categories != nil {
		next.Categories, err = normalizeValues("categories", *input.Categories, "", 9, 60)
		if err != nil {
			return Item{}, err
		}
		if len(next.Categories) == 0 {
			return Item{}, ValidationError{Field: "categories", Message: "choose at least one category"}
		}
	}
	if input.Location != nil {
		next.Location = fallback(strings.TrimSpace(*input.Location), "Unassigned")
		if utf8.RuneCountInString(next.Location) > 80 {
			return Item{}, ValidationError{Field: "location", Message: "must be 80 characters or fewer"}
		}
	}
	if input.Unit != nil {
		next.Unit = fallback(strings.TrimSpace(*input.Unit), "item")
		if utf8.RuneCountInString(next.Unit) > 30 {
			return Item{}, ValidationError{Field: "unit", Message: "must be 30 characters or fewer"}
		}
	}
	if input.MinQuantity != nil {
		if *input.MinQuantity < 0 {
			return Item{}, ValidationError{Field: "minQuantity", Message: "must be zero or greater"}
		}
		next.MinQuantity = *input.MinQuantity
	}
	if len(next.Categories) == 0 {
		next.Categories = []string{fallback(strings.TrimSpace(next.Category), "Other")}
	}
	next.Category = next.Categories[0]
	now := s.now()
	next.UpdatedAt = now
	updated, err := s.repository.UpdateItem(ctx, next)
	if err != nil {
		return Item{}, err
	}
	events, err := s.repository.ListEvents(ctx, updated.ID, now.Add(-90*24*time.Hour))
	if err != nil {
		return Item{}, err
	}
	enrichItem(&updated, events, now)
	return updated, nil
}

func (s *Service) UpdateItemMetadata(ctx context.Context, itemID string, input UpdateItemMetadataInput) (Item, error) {
	return s.UpdateItem(ctx, itemID, input)
}

func (s *Service) ListEvents(ctx context.Context, itemID string, since time.Time) ([]StockEvent, error) {
	itemID = strings.TrimSpace(itemID)
	if _, err := s.repository.GetItem(ctx, itemID); err != nil {
		return nil, err
	}
	return s.repository.ListEvents(ctx, itemID, since)
}

func (s *Service) ArchiveItem(ctx context.Context, itemID string) (Item, error) {
	s.mutationMu.Lock()
	defer s.mutationMu.Unlock()

	itemID = strings.TrimSpace(itemID)
	item, err := s.repository.GetItem(ctx, itemID)
	if err != nil {
		return Item{}, err
	}
	if item.ArchivedAt != nil {
		return s.enrich(ctx, item, s.now())
	}
	return s.archive(ctx, itemID, s.now(), true)
}

func (s *Service) RestoreItem(ctx context.Context, itemID string) (Item, error) {
	s.mutationMu.Lock()
	defer s.mutationMu.Unlock()

	itemID = strings.TrimSpace(itemID)
	item, err := s.repository.GetItem(ctx, itemID)
	if err != nil {
		return Item{}, err
	}
	if item.ArchivedAt == nil {
		return s.enrich(ctx, item, s.now())
	}
	return s.archive(ctx, itemID, s.now(), false)
}

func (s *Service) archive(ctx context.Context, itemID string, now time.Time, archived bool) (Item, error) {
	var (
		item Item
		err  error
	)
	if archived {
		item, err = s.repository.ArchiveItem(ctx, itemID, now)
	} else {
		item, err = s.repository.RestoreItem(ctx, itemID, now)
	}
	if err != nil {
		return Item{}, err
	}
	return s.enrich(ctx, item, now)
}

func (s *Service) enrich(ctx context.Context, item Item, now time.Time) (Item, error) {
	events, err := s.repository.ListEvents(ctx, item.ID, now.Add(-90*24*time.Hour))
	if err != nil {
		return Item{}, err
	}
	enrichItem(&item, events, now)
	return item, nil
}

func (s *Service) ApplyEvent(ctx context.Context, itemID string, input ApplyEventInput) (Item, error) {
	s.mutationMu.Lock()
	defer s.mutationMu.Unlock()

	if !input.Type.Valid() {
		return Item{}, ValidationError{Field: "type", Message: "must be consume, restock, or mark_level"}
	}
	note := strings.TrimSpace(input.Note)
	if utf8.RuneCountInString(note) > 240 {
		return Item{}, ValidationError{Field: "note", Message: "must be 240 characters or fewer"}
	}
	item, err := s.repository.GetItem(ctx, strings.TrimSpace(itemID))
	if err != nil {
		return Item{}, err
	}

	next := item
	eventQuantity := input.Quantity
	switch input.Type {
	case EventConsume:
		if item.TrackingMode == TrackingExact {
			if input.Quantity <= 0 {
				return Item{}, ValidationError{Field: "quantity", Message: "must be greater than zero"}
			}
			eventQuantity = math.Min(input.Quantity, item.Quantity)
			next.Quantity = item.Quantity - eventQuantity
			next.StockLevel = exactLevel(next.Quantity, item.MinQuantity)
		} else {
			points := input.Quantity
			if points == 0 {
				points = 25
			}
			if points < 0 {
				return Item{}, ValidationError{Field: "quantity", Message: "must be zero or greater"}
			}
			eventQuantity = points
			next.LevelPercent = math.Max(0, item.LevelPercent-points)
			next.StockLevel = simpleLevel(next.LevelPercent)
		}
	case EventRestock:
		if item.TrackingMode == TrackingExact {
			if input.Quantity <= 0 {
				return Item{}, ValidationError{Field: "quantity", Message: "must be greater than zero"}
			}
			next.Quantity += input.Quantity
			next.StockLevel = exactLevel(next.Quantity, next.MinQuantity)
		} else {
			next.LevelPercent = 100
			next.StockLevel = StockFull
		}
	case EventMarkLevel:
		if item.TrackingMode == TrackingSimple {
			if input.LevelPercent == nil || !validPercent(*input.LevelPercent) {
				return Item{}, ValidationError{Field: "levelPercent", Message: "must be between zero and 100"}
			}
			next.LevelPercent = *input.LevelPercent
			next.StockLevel = simpleLevel(next.LevelPercent)
		} else {
			if !input.StockLevel.Valid() {
				return Item{}, ValidationError{Field: "stockLevel", Message: "must be full, okay, low, or out"}
			}
			next.StockLevel = input.StockLevel
			if input.StockLevel == StockOut {
				next.Quantity = 0
			}
		}
	}

	now := s.now()
	next.UpdatedAt = now
	event := StockEvent{
		ID:           s.newID(),
		ItemID:       item.ID,
		Type:         input.Type,
		Quantity:     eventQuantity,
		StockLevel:   next.StockLevel,
		LevelPercent: next.LevelPercent,
		Note:         note,
		OccurredAt:   now,
	}
	updated, err := s.repository.ApplyEvent(ctx, event, next)
	if err != nil {
		return Item{}, err
	}
	events, err := s.repository.ListEvents(ctx, item.ID, now.Add(-90*24*time.Hour))
	if err != nil {
		return Item{}, err
	}
	enrichItem(&updated, events, now)
	return updated, nil
}

func enrichItem(item *Item, events []StockEvent, now time.Time) {
	item.Forecast = CalculateForecast(*item, events, now)
	item.Cadence = CalculateCadence(events, now)
}

func CalculateForecast(item Item, events []StockEvent, now time.Time) *Forecast {
	if item.TrackingMode != TrackingExact || item.Quantity <= 0 {
		return nil
	}
	var total float64
	var earliest time.Time
	count := 0
	for _, event := range events {
		if event.Type != EventConsume || event.Quantity <= 0 || event.OccurredAt.After(now) {
			continue
		}
		total += event.Quantity
		count++
		if earliest.IsZero() || event.OccurredAt.Before(earliest) {
			earliest = event.OccurredAt
		}
	}
	if count < 2 || total <= 0 {
		return nil
	}
	days := now.Sub(earliest).Hours() / 24
	if days < 1 {
		days = 1
	}
	dailyUsage := total / days
	if dailyUsage <= 0 {
		return nil
	}
	return &Forecast{
		DailyUsage:    roundOne(dailyUsage),
		DaysRemaining: roundOne(item.Quantity / dailyUsage),
		Confidence:    confidenceFor(count),
	}
}

func CalculateCadence(events []StockEvent, now time.Time) *Cadence {
	timestamps := make([]time.Time, 0, len(events))
	for _, event := range events {
		if event.Type == EventConsume && !event.OccurredAt.After(now) {
			timestamps = append(timestamps, event.OccurredAt)
		}
	}
	if len(timestamps) < 2 {
		return nil
	}
	sort.Slice(timestamps, func(left, right int) bool { return timestamps[left].Before(timestamps[right]) })
	spanDays := timestamps[len(timestamps)-1].Sub(timestamps[0]).Hours() / 24
	if spanDays <= 0 {
		return nil
	}
	averageInterval := spanDays / float64(len(timestamps)-1)
	return &Cadence{
		AverageIntervalDays: roundOne(averageInterval),
		EventsPerWeek:       roundOne(7 / averageInterval),
		LastConsumedAt:      timestamps[len(timestamps)-1],
		Confidence:          confidenceFor(len(timestamps)),
	}
}

func exactLevel(quantity, minimum float64) StockLevel {
	if quantity <= 0 {
		return StockOut
	}
	if quantity <= minimum {
		return StockLow
	}
	return StockOkay
}

func simpleLevel(percent float64) StockLevel {
	switch {
	case percent <= 0:
		return StockOut
	case percent <= 25:
		return StockLow
	case percent <= 75:
		return StockOkay
	default:
		return StockFull
	}
}

func validPercent(percent float64) bool { return percent >= 0 && percent <= 100 }

func percentForLevel(level StockLevel) float64 {
	switch level {
	case StockFull:
		return 100
	case StockLow:
		return 25
	case StockOut:
		return 0
	default:
		return 50
	}
}

func confidenceFor(count int) Confidence {
	if count >= 8 {
		return ConfidenceHigh
	}
	if count >= 4 {
		return ConfidenceMedium
	}
	return ConfidenceLow
}

func normalizeValues(field string, values []string, excluded string, maxCount, maxLength int) ([]string, error) {
	result := make([]string, 0, len(values))
	for _, value := range values {
		trimmed := strings.TrimSpace(value)
		if trimmed == "" || (excluded != "" && strings.EqualFold(trimmed, excluded)) {
			continue
		}
		if utf8.RuneCountInString(trimmed) > maxLength {
			return nil, ValidationError{Field: field, Message: fmt.Sprintf("each value must be %d characters or fewer", maxLength)}
		}
		duplicate := false
		for _, existing := range result {
			if strings.EqualFold(existing, trimmed) {
				duplicate = true
				break
			}
		}
		if !duplicate {
			result = append(result, trimmed)
		}
	}
	if len(result) > maxCount {
		return nil, ValidationError{Field: field, Message: fmt.Sprintf("use no more than %d values", maxCount)}
	}
	return result, nil
}

func fallback(value, defaultValue string) string {
	if value == "" {
		return defaultValue
	}
	return value
}

func roundOne(value float64) float64 { return math.Round(value*10) / 10 }

func randomID() string {
	bytes := make([]byte, 12)
	if _, err := rand.Read(bytes); err != nil {
		return time.Now().UTC().Format("20060102150405.000000000")
	}
	return hex.EncodeToString(bytes)
}

File: apps/worker/src/inventory/service.ts
import type { CommandContext } from "../platform/context";
import { ValidationError } from "../platform/errors";
import { createId } from "../platform/ids";
import { calculateCadence, calculateForecast } from "./forecast";
import type {
  ApplyStockEventInput,
  CreateItemInput,
  InventoryFilter,
  InventoryItem,
  StockEvent,
  UpdateItemInput,
} from "./model";
import type { AuditChange, InventoryMutation, InventoryRepository } from "./repository";
import {
  assertLength,
  assertNonNegative,
  assertPercentage,
  assertPositive,
  exactStockLevel,
  isStockLevel,
  normalizeValues,
  percentageForLevel,
  simpleStockLevel,
  trimmed,
  withDefault,
} from "./validation";

interface InventoryServiceOptions {
  now?: () => Date;
  newId?: () => string;
}

export class InventoryService {
  private readonly now: () => Date;
  private readonly newId: () => string;

  constructor(
    private readonly repository: InventoryRepository,
    options: InventoryServiceOptions = {},
  ) {
    this.now = options.now ?? (() => new Date());
    this.newId = options.newId ?? createId;
  }

  async listItems(householdId: string, filter: InventoryFilter = {}): Promise<InventoryItem[]> {
    const items = await this.repository.listItems(householdId, {
      ...filter,
      archived: filter.archived ?? "exclude",
    });
    return Promise.all(items.map((item) => this.enrich(item)));
  }

  async getItem(householdId: string, itemId: string): Promise<InventoryItem> {
    return this.enrich(await this.repository.getItem(householdId, itemId.trim()));
  }

  async createItem(context: CommandContext, input: CreateItemInput): Promise<InventoryItem> {
    const name = trimmed(input.name);
    if (!name) throw new ValidationError("name", "enter an item name");
    assertLength("name", name, 120);

    const quantity = input.quantity ?? 0;
    const minQuantity = input.minQuantity ?? 0;
    assertNonNegative("quantity", quantity);
    assertNonNegative("minQuantity", minQuantity);

    const alternativeNames = normalizeValues("alternativeNames", input.alternativeNames ?? [], {
      excluded: name,
      maximumCount: 8,
      maximumLength: 120,
    });
    const categoryInput = input.categories?.length
      ? input.categories
      : input.category
        ? [input.category]
        : ["Other"];
    const categories = normalizeValues("categories", categoryInput, {
      maximumCount: 9,
      maximumLength: 60,
    });
    if (categories.length === 0) categories.push("Other");

    const trackingMode = input.trackingMode ?? "simple";
    if (trackingMode !== "simple" && trackingMode !== "exact") {
      throw new ValidationError("trackingMode", "must be simple or exact");
    }

    let levelPercent = 0;
    let stockLevel;
    if (trackingMode === "exact") {
      stockLevel = exactStockLevel(quantity, minQuantity);
    } else {
      if (input.levelPercent !== undefined) {
        assertPercentage(input.levelPercent);
        levelPercent = input.levelPercent;
      } else if (isStockLevel(input.stockLevel)) {
        levelPercent = percentageForLevel(input.stockLevel);
      } else {
        levelPercent = 50;
      }
      stockLevel = simpleStockLevel(levelPercent);
    }

    const location = withDefault(input.location, "Unassigned");
    const unit = withDefault(input.unit, "item");
    assertLength("location", location, 80);
    assertLength("unit", unit, 30);
    const timestamp = this.now().toISOString();
    const item: InventoryItem = {
      id: this.newId(),
      householdId: context.householdId,
      name,
      alternativeNames,
      category: categories[0]!,
      categories,
      location,
      unit,
      trackingMode,
      quantity,
      stockLevel,
      levelPercent,
      minQuantity,
      version: 1,
      archivedAt: null,
      createdAt: timestamp,
      updatedAt: timestamp,
    };

    return this.repository.createItem(item, {
      context,
      action: "inventory.item.created",
      expectedVersion: 0,
      changes: createdFields(item),
    });
  }

  async updateItem(context: CommandContext, itemId: string, input: UpdateItemInput): Promise<InventoryItem> {
    const current = await this.repository.getItem(context.householdId, itemId.trim());
    const next = structuredClone(current);

    if (input.name !== undefined) {
      next.name = trimmed(input.name);
      if (!next.name) throw new ValidationError("name", "enter an item name");
      assertLength("name", next.name, 120);
    }
    if (input.alternativeNames !== undefined) {
      next.alternativeNames = normalizeValues("alternativeNames", input.alternativeNames, {
        excluded: next.name,
        maximumCount: 8,
        maximumLength: 120,
      });
    } else if (input.name !== undefined) {
      next.alternativeNames = normalizeValues("alternativeNames", next.alternativeNames, {
        excluded: next.name,
        maximumCount: 8,
        maximumLength: 120,
      });
    }
    if (input.categories !== undefined) {
      next.categories = normalizeValues("categories", input.categories, {
        maximumCount: 9,
        maximumLength: 60,
      });
      if (next.categories.length === 0) {
        throw new ValidationError("categories", "choose at least one category");
      }
      next.category = next.categories[0]!;
    }
    if (input.location !== undefined) {
      next.location = withDefault(input.location, "Unassigned");
      assertLength("location", next.location, 80);
    }
    if (input.unit !== undefined) {
      next.unit = withDefault(input.unit, "item");
      assertLength("unit", next.unit, 30);
    }
    if (input.minQuantity !== undefined) {
      assertNonNegative("minQuantity", input.minQuantity);
      next.minQuantity = input.minQuantity;
    }

    next.updatedAt = this.now().toISOString();
    next.version = current.version + 1;
    const changes = changedFields(current, next, [
      "name",
      "alternativeNames",
      "categories",
      "location",
      "unit",
      "minQuantity",
    ]);
    const updated = await this.repository.updateItem(
      next,
      this.mutation(context, "inventory.item.updated", current.version, changes),
    );
    return this.enrich(updated);
  }

  async applyStockEvent(
    context: CommandContext,
    itemId: string,
    input: ApplyStockEventInput,
  ): Promise<InventoryItem> {
    if (input.type !== "consume" && input.type !== "restock" && input.type !== "mark_level") {
      throw new ValidationError("type", "must be consume, restock, or mark_level");
    }
    const note = trimmed(input.note);
    assertLength("note", note, 240);
    const current = await this.repository.getItem(context.householdId, itemId.trim());
    const next = structuredClone(current);
    let eventQuantity = input.quantity ?? 0;

    if (input.type === "consume") {
      if (current.trackingMode === "exact") {
        assertPositive("quantity", eventQuantity);
        eventQuantity = Math.min(eventQuantity, current.quantity);
        next.quantity = current.quantity - eventQuantity;
        next.stockLevel = exactStockLevel(next.quantity, current.minQuantity);
      } else {
        if (eventQuantity === 0) eventQuantity = 25;
        assertNonNegative("quantity", eventQuantity);
        next.levelPercent = Math.max(0, current.levelPercent - eventQuantity);
        next.stockLevel = simpleStockLevel(next.levelPercent);
      }
    } else if (input.type === "restock") {
      if (current.trackingMode === "exact") {
        assertPositive("quantity", eventQuantity);
        next.quantity = current.quantity + eventQuantity;
        next.stockLevel = exactStockLevel(next.quantity, current.minQuantity);
      } else {
        next.levelPercent = 100;
        next.stockLevel = "full";
      }
    } else if (current.trackingMode === "simple") {
      assertPercentage(input.levelPercent);
      next.levelPercent = input.levelPercent;
      next.stockLevel = simpleStockLevel(input.levelPercent);
    } else {
      if (!isStockLevel(input.stockLevel)) {
        throw new ValidationError("stockLevel", "must be full, okay, low, or out");
      }
      next.stockLevel = input.stockLevel;
      if (input.stockLevel === "out") next.quantity = 0;
    }

    const timestamp = this.now().toISOString();
    next.updatedAt = timestamp;
    next.version = current.version + 1;
    const event: StockEvent = {
      id: this.newId(),
      householdId: context.householdId,
      itemId: current.id,
      type: input.type,
      quantity: eventQuantity,
      stockLevel: next.stockLevel,
      levelPercent: next.levelPercent,
      note,
      actorId: context.actorId,
      occurredAt: timestamp,
      createdAt: timestamp,
    };
    const changes = changedFields(current, next, ["quantity", "stockLevel", "levelPercent"]);
    const updated = await this.repository.applyStockEvent(
      next,
      event,
      this.mutation(context, `inventory.stock.${input.type}`, current.version, changes),
    );
    return this.enrich(updated);
  }

  async archiveItem(context: CommandContext, itemId: string): Promise<InventoryItem> {
    return this.setArchived(context, itemId, true);
  }

  async restoreItem(context: CommandContext, itemId: string): Promise<InventoryItem> {
    return this.setArchived(context, itemId, false);
  }

  async listEvents(householdId: string, itemId: string, since = new Date(0).toISOString()): Promise<StockEvent[]> {
    await this.repository.getItem(householdId, itemId.trim());
    return this.repository.listEvents(householdId, itemId.trim(), since);
  }

  private async setArchived(context: CommandContext, itemId: string, archived: boolean): Promise<InventoryItem> {
    const current = await this.repository.getItem(context.householdId, itemId.trim());
    if ((current.archivedAt !== null) === archived) return this.enrich(current);

    const next = structuredClone(current);
    next.archivedAt = archived ? this.now().toISOString() : null;
    next.updatedAt = this.now().toISOString();
    next.version = current.version + 1;
    const updated = await this.repository.setArchived(
      next,
      this.mutation(
        context,
        archived ? "inventory.item.archived" : "inventory.item.restored",
        current.version,
        changedFields(current, next, ["archivedAt"]),
      ),
    );
    return this.enrich(updated);
  }

  private mutation(
    context: CommandContext,
    action: string,
    currentVersion: number,
    changes: AuditChange[],
  ): InventoryMutation {
    return {
      context,
      action,
      expectedVersion: context.expectedVersion ?? currentVersion,
      changes,
    };
  }

  private async enrich(item: InventoryItem): Promise<InventoryItem> {
    const now = this.now();
    const since = new Date(now.getTime() - 90 * 24 * 60 * 60 * 1000).toISOString();
    const events = await this.repository.listEvents(item.householdId, item.id, since);
    const enriched = structuredClone(item);
    const forecast = calculateForecast(enriched, events, now);
    const cadence = calculateCadence(events, now);
    if (forecast) enriched.forecast = forecast;
    else delete enriched.forecast;
    if (cadence) enriched.cadence = cadence;
    else delete enriched.cadence;
    return enriched;
  }
}

function changedFields(
  before: InventoryItem,
  after: InventoryItem,
  fields: Array<keyof InventoryItem>,
): AuditChange[] {
  const changes: AuditChange[] = [];
  for (const field of fields) {
    if (JSON.stringify(before[field]) === JSON.stringify(after[field])) continue;
    changes.push({
      op: "replace",
      path: `/${field}`,
      oldValue: before[field],
      value: after[field],
    });
  }
  return changes;
}

function createdFields(item: InventoryItem): AuditChange[] {
  const fields: Array<keyof InventoryItem> = [
    "name",
    "alternativeNames",
    "categories",
    "location",
    "unit",
    "trackingMode",
    "quantity",
    "stockLevel",
    "levelPercent",
    "minQuantity",
  ];
  return fields.map((field) => ({ op: "add", path: `/${field}`, value: item[field] }));
}
